File: platform/polyglot/github_api/mjs/src/routes/repos.mjs
/**
 * Repository routes.
 * Exposes GitHub Repository API operations as REST endpoints.
 * @module routes/repos
 */

/**
 * Fastify plugin that registers repository routes.
 * @param {import('fastify').FastifyInstance} fastify
 * @param {Object} opts
 * @param {import('../sdk/repos/client.mjs').ReposClient} opts.repos - Repos SDK client
 */
export default async function repoRoutes(fastify, opts) {
  const { repos } = opts;

  /**
   * GET /repos/:owner/:repo - Get a repository.
   */
  fastify.get('/repos/:owner/:repo', async (request, reply) => {
    const { owner, repo } = request.params;
    const result = await repos.get(owner, repo);
    return reply.send(result);
  });

  /**
   * GET /repos/user/:username - List repositories for a user.
   */
  fastify.get('/repos/user/:username', async (request, reply) => {
    const { username } = request.params;
    const result = await repos.listForUser(username, request.query);
    return reply.send(result);
  });

  /**
   * GET /repos/me - List repositories for the authenticated user.
   */
  fastify.get('/repos/me', async (request, reply) => {
    const result = await repos.listForAuthenticatedUser(request.query);
    return reply.send(result);
  });

  /**
   * GET /repos/org/:org - List repositories for an organization.
   */
  fastify.get('/repos/org/:org', async (request, reply) => {
    const { org } = request.params;
    const result = await repos.listForOrg(org, request.query);
    return reply.send(result);
  });

  /**
   * POST /repos - Create a repository for the authenticated user.
   */
  fastify.post('/repos', async (request, reply) => {
    const result = await repos.create(request.body);
    return reply.status(201).send(result);
  });

  /**
   * POST /repos/org/:org - Create a repository in an organization.
   */
  fastify.post('/repos/org/:org', async (request, reply) => {
    const { org } = request.params;
    const result = await repos.createInOrg(org, request.body);
    return reply.status(201).send(result);
  });

  /**
   * PATCH /repos/:owner/:repo - Update a repository.
   */
  fastify.patch('/repos/:owner/:repo', async (request, reply) => {
    const { owner, repo } = request.params;
    const result = await repos.update(owner, repo, request.body);
    return reply.send(result);
  });

  /**
   * DELETE /repos/:owner/:repo - Delete a repository.
   */
  fastify.delete('/repos/:owner/:repo', async (request, reply) => {
    const { owner, repo } = request.params;
    await repos.delete(owner, repo);
    return reply.status(204).send();
  });

  /**
   * GET /repos/:owner/:repo/topics - Get repository topics.
   */
  fastify.get('/repos/:owner/:repo/topics', async (request, reply) => {
    const { owner, repo } = request.params;
    const result = await repos.getTopics(owner, repo);
    return reply.send(result);
  });

  /**
   * PUT /repos/:owner/:repo/topics - Replace repository topics.
   */
  fastify.put('/repos/:owner/:repo/topics', async (request, reply) => {
    const { owner, repo } = request.params;
    const { names } = request.body;
    const result = await repos.replaceTopics(owner, repo, names);
    return reply.send(result);
  });

  /**
   * GET /repos/:owner/:repo/languages - Get repository languages.
   */
  fastify.get('/repos/:owner/:repo/languages', async (request, reply) => {
    const { owner, repo } = request.params;
    const result = await repos.getLanguages(owner, repo);
    return reply.send(result);
  });

  /**
   * GET /repos/:owner/:repo/contributors - List repository contributors.
   */
  fastify.get('/repos/:owner/:repo/contributors', async (request, reply) => {
    const { owner, repo } = request.params;
    const result = await repos.listContributors(owner, repo, request.query);
    return reply.send(result);
  });

  /**
   * POST /repos/:owner/:repo/forks - Fork a repository.
   */
  fastify.post('/repos/:owner/:repo/forks', async (request, reply) => {
    const { owner, repo } = request.params;
    const result = await repos.fork(owner, repo, request.body || {});
    return reply.status(202).send(result);
  });

  /**
   * GET /repos/:owner/:repo/forks - List forks.
   */
  fastify.get('/repos/:owner/:repo/forks', async (request, reply) => {
    const { owner, repo } = request.params;
    const result = await repos.listForks(owner, repo, request.query);
    return reply.send(result);
  });

  /**
   * PUT /repos/:owner/:repo/subscription - Watch a repository.
   */
  fastify.put('/repos/:owner/:repo/subscription', async (request, reply) => {
    const { owner, repo } = request.params;
    const result = await repos.watch(owner, repo);
    return reply.send(result);
  });

  /**
   * DELETE /repos/:owner/:repo/subscription - Unwatch a repository.
   */
  fastify.delete(
    '/repos/:owner/:repo/subscription',
    async (request, reply) => {
      const { owner, repo } = request.params;
      await repos.unwatch(owner, repo);
      return reply.status(204).send();
    },
  );

  /**
   * GET /repos/:owner/:repo/commits - List commits.
   */
  fastify.get('/repos/:owner/:repo/commits', async (request, reply) => {
    const { owner, repo } = request.params;
    const result = await repos.getCommits(owner, repo, request.query);
    return reply.send(result);
  });

  /**
   * GET /repos/:owner/:repo/commits/:ref - Get a single commit.
   */
  fastify.get('/repos/:owner/:repo/commits/:ref', async (request, reply) => {
    const { owner, repo, ref } = request.params;
    const result = await repos.getCommit(owner, repo, ref);
    return reply.send(result);
  });

  /**
   * GET /repos/:owner/:repo/commits/:ref/pulls - List PRs associated with a commit.
   */
  fastify.get('/repos/:owner/:repo/commits/:ref/pulls', async (request, reply) => {
    const { owner, repo, ref } = request.params;
    const result = await repos.listCommitPulls(owner, repo, ref);
    return reply.send(result);
  });

  /**
   * GET /repos/:owner/:repo/contents - Get root contents.
   */
  fastify.get('/repos/:owner/:repo/contents', async (request, reply) => {
    const { owner, repo } = request.params;
    const result = await repos.getContents(owner, repo, '', request.query);
    return reply.send(result);
  });

  /**
   * GET /repos/:owner/:repo/contents/* - Get contents at path.
   */
  fastify.get('/repos/:owner/:repo/contents/*', async (request, reply) => {
    const { owner, repo } = request.params;
    const path = request.params['*'];
    const result = await repos.getContents(owner, repo, path, request.query);
    return reply.send(result);
  });

  /**
   * GET /repos/:owner/:repo/git/trees/:tree_sha - Get a git tree.
   */
  fastify.get('/repos/:owner/:repo/git/trees/:tree_sha', async (request, reply) => {
    const { owner, repo, tree_sha } = request.params;
    const result = await repos.getGitTree(owner, repo, tree_sha, request.query);
    return reply.send(result);
  });

  /**
   * PUT /repos/:owner/:repo/contents/* - Create or update file contents.
   */
  fastify.put('/repos/:owner/:repo/contents/*', async (request, reply) => {
    const { owner, repo } = request.params;
    const path = request.params['*'];
    const result = await repos.createOrUpdateContents(owner, repo, path, request.body);
    return reply.status(201).send(result);
  });

  /**
   * POST /repos/:owner/:repo/git/refs - Create a git reference (branch/tag).
   */
  fastify.post('/repos/:owner/:repo/git/refs', async (request, reply) => {
    const { owner, repo } = request.params;
    const result = await repos.createGitRef(owner, repo, request.body.ref, request.body.sha);
    return reply.status(201).send(result);
  });

  /**
   * POST /repos/:owner/:repo/pulls - Create a pull request.
   */
  fastify.post('/repos/:owner/:repo/pulls', async (request, reply) => {
    const { owner, repo } = request.params;
    const result = await repos.createPullRequest(owner, repo, request.body);
    return reply.status(201).send(result);
  });

  /**
   * POST /repos/:owner/:repo/issues/:issueNumber/labels - Add labels to an issue or PR.
   */
  fastify.post('/repos/:owner/:repo/issues/:issueNumber/labels', async (request, reply) => {
    const { owner, repo, issueNumber } = request.params;
    const { labels } = request.body || {};
    const result = await repos.addLabels(owner, repo, parseInt(issueNumber, 10), labels);
    return reply.status(200).send(result);
  });

  return Promise.resolve();
}

File: platform/polyglot/github_api/py/github_api/routes/repos.py
"""
Repository routes for the FastAPI server.

Maps HTTP endpoints to ReposClient methods with request/response handling.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Request

__all__ = ["router"]

router = APIRouter(tags=["repos"])


def _get_repos_client(request: Request):
    """Get the ReposClient from app state."""
    from github_api.sdk.repos import ReposClient
    return ReposClient(request.app.state.github_client)


# ── Static/specific routes MUST be registered before parametric /{owner}/{repo}
# ── FastAPI matches routes in registration order (unlike Fastify's radix tree
# ── which automatically prioritises static segments over parametric ones).


@router.get("/repos/me")
async def list_my_repos(request: Request) -> dict[str, Any]:
    """List repositories for the authenticated user."""
    client = _get_repos_client(request)
    params = dict(request.query_params)
    return await client.list_for_authenticated_user(params=params or None)


@router.get("/repos/user/{username}")
async def list_user_repos(request: Request, username: str) -> dict[str, Any]:
    """List repositories for a user."""
    client = _get_repos_client(request)
    params = dict(request.query_params)
    return await client.list_for_user(username, params=params or None)


@router.get("/repos/org/{org}")
async def list_org_repos(request: Request, org: str) -> dict[str, Any]:
    """List repositories for an organization."""
    client = _get_repos_client(request)
    params = dict(request.query_params)
    return await client.list_for_org(org, params=params or None)


@router.post("/repos")
async def create_repo(request: Request) -> dict[str, Any]:
    """Create a repository for the authenticated user."""
    client = _get_repos_client(request)
    body = await request.json()
    return await client.create(body)


@router.post("/repos/org/{org}")
async def create_org_repo(request: Request, org: str) -> dict[str, Any]:
    """Create a repository in an organization."""
    client = _get_repos_client(request)
    body = await request.json()
    return await client.create_in_org(org, body)


@router.get("/repos/{owner}/{repo}")
async def get_repository(request: Request, owner: str, repo: str) -> dict[str, Any]:
    """Get a repository by owner and name."""
    client = _get_repos_client(request)
    return await client.get(owner, repo)


@router.patch("/repos/{owner}/{repo}")
async def update_repo(request: Request, owner: str, repo: str) -> dict[str, Any]:
    """Update a repository."""
    client = _get_repos_client(request)
    body = await request.json()
    return await client.update(owner, repo, body)


@router.delete("/repos/{owner}/{repo}")
async def delete_repo(request: Request, owner: str, repo: str) -> dict[str, Any]:
    """Delete a repository."""
    client = _get_repos_client(request)
    return await client.delete(owner, repo)


@router.get("/repos/{owner}/{repo}/topics")
async def get_topics(request: Request, owner: str, repo: str) -> dict[str, Any]:
    """Get repository topics."""
    client = _get_repos_client(request)
    return await client.get_topics(owner, repo)


@router.put("/repos/{owner}/{repo}/topics")
async def replace_topics(request: Request, owner: str, repo: str) -> dict[str, Any]:
    """Replace all repository topics."""
    client = _get_repos_client(request)
    body = await request.json()
    return await client.replace_topics(owner, repo, body.get("names", []))


@router.get("/repos/{owner}/{repo}/languages")
async def get_languages(request: Request, owner: str, repo: str) -> dict[str, Any]:
    """Get repository language breakdown."""
    client = _get_repos_client(request)
    return await client.get_languages(owner, repo)


@router.get("/repos/{owner}/{repo}/contributors")
async def list_contributors(request: Request, owner: str, repo: str) -> dict[str, Any]:
    """List repository contributors."""
    client = _get_repos_client(request)
    params = dict(request.query_params)
    return await client.list_contributors(owner, repo, params=params or None)


@router.post("/repos/{owner}/{repo}/forks")
async def create_fork(request: Request, owner: str, repo: str) -> dict[str, Any]:
    """Create a fork of a repository."""
    client = _get_repos_client(request)
    try:
        body = await request.json()
    except Exception:
        body = None
    return await client.fork(owner, repo, options=body)


@router.get("/repos/{owner}/{repo}/forks")
async def list_forks(request: Request, owner: str, repo: str) -> dict[str, Any]:
    """List forks of a repository."""
    client = _get_repos_client(request)
    params = dict(request.query_params)
    return await client.list_forks(owner, repo, params=params or None)


@router.put("/repos/{owner}/{repo}/subscription")
async def watch_repo(request: Request, owner: str, repo: str) -> dict[str, Any]:
    """Watch (subscribe to) a repository."""
    client = _get_repos_client(request)
    return await client.watch(owner, repo)


@router.delete("/repos/{owner}/{repo}/subscription")
async def unwatch_repo(request: Request, owner: str, repo: str) -> dict[str, Any]:
    """Unwatch (unsubscribe from) a repository."""
    client = _get_repos_client(request)
    return await client.unwatch(owner, repo)


@router.get("/repos/{owner}/{repo}/commits")
async def list_commits(request: Request, owner: str, repo: str):
    """List commits for a repository."""
    client = _get_repos_client(request)
    params = dict(request.query_params)
    return await client.get_commits(owner, repo, params=params or None)


@router.get("/repos/{owner}/{repo}/commits/{ref}")
async def get_commit(request: Request, owner: str, repo: str, ref: str):
    """Get a single commit by ref."""
    client = _get_repos_client(request)
    return await client.get_commit(owner, repo, ref)


@router.get("/repos/{owner}/{repo}/commits/{ref}/pulls")
async def list_commit_pulls(request: Request, owner: str, repo: str, ref: str):
    """List pull requests associated with a commit."""
    client = _get_repos_client(request)
    return await client.list_commit_pulls(owner, repo, ref)


@router.get("/repos/{owner}/{repo}/contents")
async def get_root_contents(request: Request, owner: str, repo: str):
    """Get repository root contents."""
    client = _get_repos_client(request)
    params = dict(request.query_params)
    return await client.get_contents(owner, repo, "", params=params or None)


@router.get("/repos/{owner}/{repo}/contents/{path:path}")
async def get_contents_at_path(request: Request, owner: str, repo: str, path: str):
    """Get repository contents at a specific path."""
    client = _get_repos_client(request)
    params = dict(request.query_params)
    return await client.get_contents(owner, repo, path, params=params or None)


@router.get("/repos/{owner}/{repo}/git/trees/{tree_sha}")
async def get_git_tree(request: Request, owner: str, repo: str, tree_sha: str):
    """Get a git tree by SHA or branch name."""
    client = _get_repos_client(request)
    params = dict(request.query_params)
    return await client.get_git_tree(owner, repo, tree_sha, params=params or None)
